### PR TITLE
Train NLU pipeline using chunks

### PR DESCRIPTION
**Pull request by tabergma**
**Proposed changes**:
closes https://github.com/RasaHQ/rasa/issues/7394

- Add a new data generator that loads data from files
- Update the `train_chunk` interface for classifiers and entity extractors 
- Update the train loop, training NLU models in chunks is working

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Review by dakshvar22 — COMMENTED** [review]
Looks good! ✨  Sorry for the delay in review.
Should we add a test which checks that components are trained in the correct order for different configuration of pipelines? For example, `train` for `DIET` is not called before featurizers are trained on all chunks.

File: rasa/train.py
Comment (tabergma) [reply rasa/train.py:60]: Yeah, also recognized that and undid the change. I guess I forgot some parts.

File: rasa/shared/nlu/training_data/training_data.py
Comment (tabergma) [reply rasa/shared/nlu/training_data/training_data.py:862]: for sparse features our data are always integer

Comment (dakshvar22) [reply rasa/shared/nlu/training_data/training_data.py:862]: Ahh right, I missed the if else construct above.

File: rasa/nlu/model.py
Comment (tabergma) [reply rasa/nlu/model.py:299]: We currently also inform the user about the status of the training. E.g. for each component we log 
"started training" and "finished training". Don't want to log this for every featurizer as they are called multiple times, so I just inform the user that a particular chunk is featurized.

Comment (dakshvar22) [reply rasa/nlu/model.py:299]: Right, I see your point 👍

File: rasa/nlu/classifiers/diet_classifier.py
Comment (joejuzl) [inline rasa/nlu/classifiers/diet_classifier.py:816]: Is this the right place to load the training data from files? I wonder if the data should be loaded before it's passed to the classifiers.

**Comment by tabergma** [comment]
@dakshvar22 @joejuzl FYI: I discovered a couple of more edge cases that we need to fix before we can merge this PR. Will try to wrap up as much as I can today and will then handover the PR to @Ghostvv. So no need to review for now in detail I think.

File: rasa/nlu/extractors/entity_synonyms.py
Comment (joejuzl) [inline rasa/nlu/extractors/entity_synonyms.py:80]: This means we are loading the same files multiple times right? As we also do this is the classifier?

File: rasa/shared/nlu/training_data/training_data.py
Comment (joejuzl) [inline rasa/shared/nlu/training_data/training_data.py:884]: Any reason this is a method?

File: rasa/shared/nlu/training_data/training_data.py
Comment (joejuzl) [inline rasa/shared/nlu/training_data/training_data.py:896]: Is this the right place for this? Where do we normally encode training data?

File: rasa/nlu/classifiers/diet_classifier.py
Comment (tabergma) [reply rasa/nlu/classifiers/diet_classifier.py:816]: Maybe some background information: We currently are not able to train on large datasets due to out of memory issues. That's why we are introducing a new command called `rasa train --number-of-chunks <n>`. This divides the training data into multiple smaller chunks and featurizes them separately. During training we need to train on all data, e.g. one epoch of training should see all training data. As the training data at this point is saved in multiple files on disk, as we cannot have everything in memory at the same time, we pass the data files to the training method. During training we load one data file at a time to make sure we are not running out of memory. This has the benefit that we can (hopefully) train on larger datasets, but has the downside that training will take longer as we need to load data multiple times from files.

So to answer your question, yes this is the right place as we cannot load all the data before training.

File: rasa/nlu/extractors/entity_synonyms.py
Comment (tabergma) [reply rasa/nlu/extractors/entity_synonyms.py:80]: yes, see comment above

File: rasa/shared/nlu/training_data/training_data.py
Comment (tabergma) [reply rasa/shared/nlu/training_data/training_data.py:896]: We did not encode the message data before. (Maybe encode is not the best word.) We need to write down the training data incl. features to files. As we also need to have the text and labels of the message, we need to "encode" the data into tf.Features.

File: rasa/nlu/extractors/entity_synonyms.py
Comment (tabergma) [reply rasa/nlu/extractors/entity_synonyms.py:80]: Ah no, did not see that this referred to entity synonyms. In this case each file is just loaded once.

**Comment by tabergma** [comment]
@Ghostvv Training the rasa init project is working again, e2e and incremental training messed up some things. I also now build up the label index mapping during preparation in DIET, so that we can actually check if the model can be trained (especially important for Response Selector). However, some things still not work as expected:
- training our e2e example bot fails
- testing does not work
It might be that some featurizers do not work as expected as the testing pipeline is exactly the same as before. I don't think I find the time to investigate those two issues further.

File: rasa/shared/nlu/training_data/training_data.py
Comment (tabergma) [inline rasa/shared/nlu/training_data/training_data.py:777]: We most likely need to add ACTION_TEXT and ACTION_NAME for e2e training.

Comment (Ghostvv) [reply rasa/shared/nlu/training_data/training_data.py:876]: @dakshvar22 do you remember why we need that list of constants?

File: rasa/nlu/utils/hugging_face/hf_transformers.py
Comment (Ghostvv) [inline rasa/nlu/utils/hugging_face/hf_transformers.py:693]: @dakshvar22 should this be `train_chunk`?

File: rasa/nlu/utils/spacy_utils.py
Comment (Ghostvv) [inline rasa/nlu/utils/spacy_utils.py:237]: @dakshvar22 it seems to me that these `docs` already contain dense features. I think it should be `train_chunk` then?

File: rasa/shared/nlu/training_data/training_data.py
Comment (Ghostvv) [reply rasa/shared/nlu/training_data/training_data.py:917]: I think adding `ACTION_TEXT, ACTION_NAME` is enough. I think we need `TOKEN_NAMES[TEXT]` for entities

File: rasa/nlu/classifiers/keyword_intent_classifier.py
Comment (Ghostvv) [reply rasa/nlu/classifiers/keyword_intent_classifier.py:52]: @dakshvar22 I think it should be done in the separate PR

File: rasa/nlu/utils/hugging_face/hf_transformers.py
Comment (dakshvar22) [reply rasa/nlu/utils/hugging_face/hf_transformers.py:693]: Hmm, I think because components like `HFTransformersNLP`, and `SpacyNLP` go from creating tokens to features inside one component, it doesn't need a separate `train_chunk` method explicitly because we are passing the complete training data [here](https://github.com/RasaHQ/rasa/pull/7551/files#diff-178f448a45faaee012886c9ac7564a55639e9ff7383e2c68ae4c1210020276daR285). But then those components should be doing the chunking themselves and featurizing chunks? These components are really an exception for the logic in the component class and break the abstraction between `train` and `train_chunk` I think.

File: rasa/nlu/utils/spacy_utils.py
Comment (dakshvar22) [reply rasa/nlu/utils/spacy_utils.py:237]: This one is also like the above point made about `HFTransformers`.

File: rasa/nlu/utils/hugging_face/hf_transformers.py
Comment (Ghostvv) [reply rasa/nlu/utils/hugging_face/hf_transformers.py:693]: but the problem is that they will create dense features for the whole training data, which will blow up in memory. I think the problem is that they are applied before tokenizers

Comment (dakshvar22) [reply rasa/nlu/utils/hugging_face/hf_transformers.py:693]: Yes, that's what I meant by these components being an exception for the logic in `component.py`

Comment (Ghostvv) [reply rasa/nlu/utils/hugging_face/hf_transformers.py:693]: let's add them to the to do list, and tackle it in the separate PR

**Comment by Ghostvv** [comment]
@joejuzl @dakshvar22 the tests are passing, could you please review?

File: examples/e2ebot/data/nlu.yml
Comment (joejuzl) [inline examples/e2ebot/data/nlu.yml:1]: why is this being deleted?

File: rasa/shared/nlu/training_data/message.py
Comment (joejuzl) [inline rasa/shared/nlu/training_data/message.py:131]:
```suggestion
        data = dict(sorted(data.items()))
```

File: rasa/shared/nlu/training_data/tokens.py
Comment (joejuzl) [inline rasa/shared/nlu/training_data/tokens.py:65]:
```suggestion
            token_dict.get("end"),
```

File: rasa/shared/nlu/training_data/tokens.py
Comment (joejuzl) [inline rasa/shared/nlu/training_data/tokens.py:66]:
```suggestion
            token_dict.get("data"),
```

File: rasa/shared/nlu/training_data/tokens.py
Comment (joejuzl) [inline rasa/shared/nlu/training_data/tokens.py:67]:
```suggestion
            token_dict.get("lemma")
```

File: rasa/shared/nlu/training_data/tokens.py
Comment (joejuzl) [inline rasa/shared/nlu/training_data/tokens.py:87]: does it makes sense to compare `text` for `<`?

**Review by joejuzl — CHANGES_REQUESTED** [review]

File: examples/e2ebot/data/nlu.yml
Comment (Ghostvv) [reply examples/e2ebot/data/nlu.yml:1]: no idea 😅

Comment (Ghostvv) [reply examples/e2ebot/data/nlu.yml:1]: I guess artifact of testing something

File: rasa/shared/nlu/training_data/tokens.py
Comment (Ghostvv) [reply rasa/shared/nlu/training_data/tokens.py:87]: hm, that's strange. Do you know the way to figure out where it is used?

Comment (Ghostvv) [reply rasa/shared/nlu/training_data/tokens.py:87]: modified to only compare start and end

Comment (joejuzl) [reply rasa/shared/nlu/training_data/tokens.py]: It will be used whenever the object is used with the infix operator `<`.
As for finding where it's used, you could pop a debugger/breakpoint in the method and run the tests and see if it catches.
My guess would be it's to do with sorting a list of these things.

Comment (joejuzl) [reply rasa/shared/nlu/training_data/tokens.py]: hmm actually is that is the case, maybe we do want to use `text` in the comparison...

Comment (Ghostvv) [reply rasa/shared/nlu/training_data/tokens.py]: yeah, I don't know it is implemented that way in `main`, let's not touch it

Comment (Ghostvv) [reply rasa/shared/nlu/training_data/tokens.py:87]: this piece of code is just copy pasted to `shared`

**Review by joejuzl — COMMENTED** [review]
The (albeit small) "engineering" changes look fine to me. Just checking though, the things that Tanja mentioned were broken:
> - training our e2e example bot fails
> - testing does not work

are now fixed?

I will wait for @dakshvar22 to give actual approval as he can review all the ml changes.

**Comment by Ghostvv** [comment]
> training our e2e example bot fails

nlu part works, it was related to the fact that core messages were not included into chunks

> testing does not work

hm, I thought she meant tests. I'll check `rasa test`

**Comment by Ghostvv** [comment]
`rasa test` and `rasa test nlu` seems to be working for me

File: rasa/shared/nlu/training_data/training_data.py
Comment (dakshvar22) [inline rasa/shared/nlu/training_data/training_data.py:107]: I think it's the opposite of what the function does.

File: rasa/shared/nlu/training_data/training_data.py
Comment (dakshvar22) [inline rasa/shared/nlu/training_data/training_data.py:452]: Use `TrainingDataFull` as return type

File: rasa/shared/nlu/training_data/training_data.py
Comment (dakshvar22) [inline rasa/shared/nlu/training_data/training_data.py:973]: Can we move this to the constants file?

File: rasa/utils/tensorflow/models.py
Comment (dakshvar22) [inline rasa/utils/tensorflow/models.py:89]: This can be removed I suppose?

File: tests/nlu/classifiers/test_diet_classifier.py
Comment (dakshvar22) [inline tests/nlu/classifiers/test_diet_classifier.py:498]:
```suggestion
```

File: tests/nlu/classifiers/test_diet_classifier.py
Comment (dakshvar22) [inline tests/nlu/classifiers/test_diet_classifier.py:506]:
```suggestion
```

**Review by dakshvar22 — COMMENTED** [review]
Some minor comments 🚀 
In addition, can you please make follow up issues for handling training in chunks for keyword intent classifier and language model featurizer. Also, can you please create a changelog already listing the changes like `TrainingData` is deprecated, etc. It will be easier to collect all the changes in the big PR.

I haven't tried training with this yet. Will try it later today. Also, since we refactored some bits of DIET and featurizer, we should run the model regression tests.

**Comment by Ghostvv** [comment]
> In addition, can you please make follow up issues for handling training in chunks for keyword intent classifier and language model featurizer.

already in the list, let's create actual issues when we plan the next steps

File: rasa/nlu/classifiers/diet_classifier.py
Comment (Ghostvv) [reply rasa/nlu/classifiers/diet_classifier.py:627]: when DIET performs only entity classification

File: rasa/nlu/classifiers/diet_classifier.py
Comment (Ghostvv) [reply rasa/nlu/classifiers/diet_classifier.py:818]: what do you mean? we train model in this method

File: rasa/nlu/classifiers/diet_classifier.py
Comment (Ghostvv) [reply rasa/nlu/classifiers/diet_classifier.py:892]: but we run validation in the end of each epoch, not batch

File: rasa/nlu/classifiers/diet_classifier.py
Comment (dakshvar22) [reply rasa/nlu/classifiers/diet_classifier.py:818]: Yes, but the third argument is `validation_data_generator` and hence I thought we can be a bit more explicit by renaming `data_generator` to `train_data_generator`.

File: rasa/nlu/classifiers/diet_classifier.py
Comment (dakshvar22) [reply rasa/nlu/classifiers/diet_classifier.py:892]: Yes, over here as well `data_generator` contains data for whole epoch and not one batch. Since we don't pass any `validation_data_generator` there won't be a validation run. All I'm saying is that we can possibly create a validation data generator as well when training on chunks and pass it to `_train_model` just like we do for usual training on line 926.

File: rasa/nlu/classifiers/diet_classifier.py
Comment (Ghostvv) [reply rasa/nlu/classifiers/diet_classifier.py:819]: aaah, I thought you suggest to rename the method itself

File: rasa/shared/nlu/training_data/training_data.py
Comment (Ghostvv) [reply rasa/shared/nlu/training_data/training_data.py:973]: moved it inside `TrainingDataChunk` class

File: rasa/utils/tensorflow/models.py
Comment (Ghostvv) [reply rasa/utils/tensorflow/models.py:89]: no, this is smth I don't understand still

File: rasa/nlu/classifiers/diet_classifier.py
Comment (Ghostvv) [reply rasa/nlu/classifiers/diet_classifier.py:893]: I suppose, but then we'll need to use a separate chunk for it. Let's do it in a separate PR, I'll add it to the list

File: rasa/nlu/featurizers/dense_featurizer/lm_featurizer.py
Comment (Ghostvv) [reply rasa/nlu/featurizers/dense_featurizer/lm_featurizer.py]: I think there is circular imports, test complains

**Comment by Ghostvv** [comment]
@dakshvar22 I answered your comments. ready to merge?

**Comment by dakshvar22** [comment]
> Also, since we refactored some bits of DIET and featurizer, we should run the model regression tests.

Do you want to do this or are you confident that the changes shouldn't bring a regression for now?

**Comment by Ghostvv** [comment]
/modeltest
```
include:
   - dataset: ["all"]
     config: ["all"]
```

**Comment by Ghostvv** [comment]
@dakshvar22 regression tests look good to me. The training on carbon-bot is unstable on majority of PRs

**Comment by dakshvar22** [comment]
Right, good to merge 👍

**Pull request merged**
